runelite-client/src/main/java/net/runelite/client/game/FreezeManager.java

```diff
@@ -0,0 +1,196 @@
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import lombok.extern.slf4j.Slf4j;
+import net.runelite.api.Actor;
+import net.runelite.api.Client;
+import net.runelite.api.FreezeInfo;
+import net.runelite.api.FreezeType;
+import net.runelite.api.NPC;
+import net.runelite.api.OverheadPrayer;
+import net.runelite.api.Player;
+import net.runelite.api.events.GameTick;
+import net.runelite.api.events.GraphicChanged;
+
+import java.util.List;
+
+@Slf4j
+public class FreezeManager
+{
+	public static final int IMMUNE_TICKS = 6;
+
+	private Client client;
+	private EventBus eventBus;
+
+	public FreezeManager(Client client, EventBus eventBus)
+	{
+		this.client = client;
+		this.eventBus = eventBus;
+
+		//register ourself with the event bus to process changes
+		this.eventBus.register(this);
+	}
+
+	@Subscribe
+	public void onGameTick(GameTick event)
+	{
+		List<NPC> npcs = client.getNpcs();
+		List<Player> players = client.getPlayers();
+
+		for (NPC npc : npcs) //process npcs before players just like the game does
+		{
+			processActor(npc);
+		}
+
+		for (Player player : players)
+		{
+			processActor(player);
+		}
+	}
+
+	@Subscribe
+	public void onGraphicChanged(GraphicChanged event)
+	{
+		Actor a = event.getActor();
+		int graphic = a.getGraphic();
+		FreezeType type = null;
+
+		if (graphic == FreezeType.ICE_RUSH.getGraphic())
+		{
+			type = FreezeType.ICE_RUSH;
+		}
+		else if (graphic == FreezeType.ICE_BURST.getGraphic())
+		{
+			type = FreezeType.ICE_BURST;
+		}
+		else if (graphic == FreezeType.ICE_BLITZ.getGraphic())
+		{
+			type = FreezeType.ICE_BLITZ;
+		}
+		else if (graphic == FreezeType.ICE_BARRAGE.getGraphic())
+		{
+			type = FreezeType.ICE_BARRAGE;
+		}
+		else if (graphic == FreezeType.BIND.getGraphic())
+		{
+			type = FreezeType.BIND;
+		}
+		else if (graphic == FreezeType.SNARE.getGraphic())
+		{
+			type = FreezeType.SNARE;
+		}
+		else if (graphic == FreezeType.ENTANGLE.getGraphic())
+		{
+			type = FreezeType.ENTANGLE;
+		}
+
+		if (type == null)
+			return;
+
+		int freezeTicks = type.getTicks();
+
+		if (a instanceof Player)
+		{
+			log.debug("Player freeze queued");
+			OverheadPrayer activeOverhead = OverheadPrayer.fromId(((Player) a).getOverheadIcon());
+
+			if (type.isHalfOnPray() && activeOverhead == OverheadPrayer.PROTECT_MAGIC)
+			{
+				freezeTicks /= 2;
+			}
+
+			a.getFreeze().queueFreeze(type, freezeTicks, false);
+		}
+		else if (a instanceof NPC)
+		{
+			log.debug("NPC freeze queued");
+			//TODO determine halving of ticks when praying
+			a.getFreeze().queueFreeze(type, freezeTicks, false);
+		}
+	}
+
+	private static void processActor(Actor subject)
+	{
+		FreezeInfo freezeInfo = subject.getFreeze();
+		boolean wasImmune = freezeInfo.isImmune();
+
+		freezeInfo.decrementAll();
+
+		//after immunity runs out reset everything (no bars)
+		if (!freezeInfo.isFrozen() && !freezeInfo.isImmune() && freezeInfo.getType() == null && wasImmune)
```

Comment:
This is a little hard to grok.

Should this be a calculating function on `freezeInfo` like `freezeInfo.noMoreImmunity() && wasImmune`?